sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py

```diff
@@ -510,35 +526,43 @@ def __init__(self, **kwargs: Any) -> None:
         self._update_lock = Lock()
         self._refresh_lock = Lock()
 
-    def refresh(self, **kwargs) -> None:
+    def refresh(self, **kwargs) -> None:  # pylint: disable=too-many-statements
         if not self._refresh_on and not self._feature_flag_refresh_enabled:
-            logging.debug("Refresh called but no refresh enabled.")
+            logger.debug("Refresh called but no refresh enabled.")
             return
         if not self._refresh_timer.needs_refresh():
-            logging.debug("Refresh called but refresh interval not elapsed.")
+            logger.debug("Refresh called but refresh interval not elapsed.")
             return
         if not self._refresh_lock.acquire(blocking=False):  # pylint: disable= consider-using-with
-            logging.debug("Refresh called but refresh already in progress.")
+            logger.debug("Refresh called but refresh already in progress.")
             return
         success = False
         need_refresh = False
         error_message = """
                         Failed to refresh configuration settings from Azure App Configuration.
                         """
         exception: Exception = RuntimeError(error_message)
+        is_failover_request = False
+        client_count = self._replica_client_manager.get_client_count()
         try:
             self._replica_client_manager.refresh_clients()
-            active_clients = self._replica_client_manager.get_active_clients()
+            self._replica_client_manager.find_active_clients()
+
+            while self._replica_client_manager.has_next_client():
+                headers = __update_correlation_context_header(
+                    kwargs.pop("headers", {}),
+                    "Watch",
+                    client_count - 1,
+                    self._feature_flag_enabled,
+                    self._feature_filter_usage,
+                    self._uses_key_vault,
+                    self._uses_load_balancing,
+                    is_failover_request,
+                )
 
-            headers = _get_headers(
-                kwargs.pop("headers", {}),
-                "Watch",
-                self._replica_client_manager.get_client_count() - 1,
-                self._feature_flag_enabled,
-                self._feature_filter_usage,
-                self._uses_key_vault,
-            )
-            for client in active_clients:
+                client = self._replica_client_manager.get_next_client()
+                if not client:
+                    return
```

Comment:
Since we may return here, we should probably update the headers only after we have confirmed we have a client.